File: libraries/ios/Sources/ChatUi/WebBridge/FlexAction.swift
//
//  FlexAction.swift
//  FlexHybridApp
//
//  Created by dvkyun on 2020/09/01.
//  Copyright © 2020 dvkyun. All rights reserved.
//

import Foundation

public class FlexAction {
    
    private let funcName: String
    private let mComponent: FlexComponent
    private var _isFinished = false
    
    public var isFinished: Bool {
        _isFinished
    }
    public var onFinished: (() -> Void)? = nil
    
    internal init (_ name: String, _ component: FlexComponent) {
        funcName = name
        mComponent = component
    }
    
    @MainActor private func pRetrun(_ response: Any?) {
        if _isFinished {
            FlexMsg.err(FlexString.ERROR7)
            return
        }
        if response is BrowserException {
            let reason = (response as! BrowserException).reason == nil ? "null" : "\"\((response as! BrowserException).reason!)\""
            mComponent.evalJS("$flex.flex.\(funcName)(false, \(reason))")
        } else if response == nil || response is Void {
            mComponent.evalJS("$flex.flex.\(funcName)(true)")
        } else {
            do {
                mComponent.evalJS("$flex.flex.\(funcName)(true, null, \(try FlexFunc.convertValue(response!)))")
            } catch FlexError.UnuseableTypeCameIn {
                FlexMsg.err(FlexString.ERROR3)
                mComponent.evalJS("$flex.flex.\(funcName)(false, \"\(FlexString.ERROR3)\")")
            } catch {
                FlexMsg.err(error)
                mComponent.evalJS("$flex.flex.\(funcName)(false, \"\(error.localizedDescription)\")")
            }
        }
        finish()
    }
    
    @MainActor public func promiseReturn(_ response: Void) {
        pRetrun(response)
    }
       
    @MainActor public func promiseReturn(_ response: String) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Int) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Float) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Double) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Character) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Bool) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Array<Any?>) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Dictionary<String,Any?>) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: BrowserException) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn(_ response: Encodable) {
        pRetrun(response)
    }
    
    @MainActor public func promiseReturn() {
        resolveVoid()
    }
    
    @MainActor public func resolveVoid() {
        if _isFinished {
            FlexMsg.err(FlexString.ERROR7)
            return
        }
        mComponent.evalJS("$flex.flex.\(funcName)(true)")
        finish()
    }
    
    @MainActor public func reject(reason: BrowserException) {
        if _isFinished {
            FlexMsg.err(FlexString.ERROR7)
            return
        }
        _isFinished = true
        let rejectReson = reason.reason == nil ? "null" : "\"\(reason.reason!)\""
        mComponent.evalJS("$flex.flex.\(funcName)(false, \(rejectReson))")
    }
    
    @MainActor public func reject(reason: String) {
        if _isFinished {
            FlexMsg.err(FlexString.ERROR7)
            return
        }
        mComponent.evalJS("$flex.flex.\(funcName)(false, \"\(reason)\")")
        finish()
    }
    
    @MainActor public func reject() {
        if _isFinished {
            FlexMsg.err(FlexString.ERROR7)
            return
        }
        mComponent.evalJS("$flex.flex.\(funcName)(false)")
        finish()
    }
    
    private func finish() {
        _isFinished = true
        onFinished?()
    }
    
}

File: libraries/android/chat-ui/src/main/java/app/dvkyun/flexhybridapp/FlexAction.kt
package app.dvkyun.flexhybridapp

import org.json.JSONArray
import org.json.JSONObject

class FlexAction internal constructor(name: String, webView: FlexWebView) {

    private var funName: String = name
    private var flexWebView: FlexWebView = webView

    var isFinished = false
        private set

    var onFinished: (() -> Unit)? = null

    private fun pReturn(response: Any?) {
        if (isFinished) {
            FlexUtil.INFO(FlexException.ERROR9)
            return
        }
        if (response is BrowserException) {
            val reason = if (response.reason == null) null else "\"${response.reason}\""
            FlexUtil.rejectPromise(flexWebView, funName, reason)
        } else if (response == null || response is Unit || response is Void) {
            FlexUtil.responsePromise(flexWebView, funName)
        } else {
            FlexUtil.responsePromise(flexWebView, funName, response)
        }
        finish()
    }

    fun promiseReturn(response: String) {
        pReturn(response)
    }

    fun promiseReturn(response: Int) {
        pReturn(response)
    }

    fun promiseReturn(response: Long) {
        pReturn(response)
    }

    fun promiseReturn(response: Float) {
        pReturn(response)
    }

    fun promiseReturn(response: Double) {
        pReturn(response)
    }

    fun promiseReturn(response: Char) {
        pReturn(response)
    }

    fun promiseReturn(response: Boolean) {
        pReturn(response)
    }

    fun promiseReturn(response: JSONArray) {
        pReturn(response)
    }

    fun promiseReturn(response: Array<*>) {
        pReturn(response)
    }

    fun promiseReturn(response: Iterable<*>) {
        pReturn(response)
    }

    fun promiseReturn(response: JSONObject) {
        pReturn(response)
    }

    fun promiseReturn(response: Map<String, *>) {
        pReturn(response)
    }

    fun <T : FlexType> promiseReturn(response: T) {
        pReturn(response)
    }

    fun promiseReturn(response: Unit) {
        pReturn(response)
    }

    fun promiseReturn(response: Void) {
        pReturn(response)
    }

    fun promiseReturn(response: BrowserException) {
        pReturn(response)
    }

    fun promiseReturn() {
        resolveVoid()
    }

    fun resolveVoid() {
        if (isFinished) {
            FlexUtil.INFO(FlexException.ERROR9)
            return
        }
        FlexUtil.responsePromise(flexWebView, funName)
        finish()
    }

    fun reject(reason: BrowserException) {
        if (isFinished) {
            FlexUtil.INFO(FlexException.ERROR9)
            return
        }
        val rejectReason = if (reason.reason == null) null else "\"${reason.reason}\""
        FlexUtil.rejectPromise(flexWebView, funName, rejectReason)
        finish()
    }

    fun reject(reason: String) {
        if (isFinished) {
            FlexUtil.INFO(FlexException.ERROR9)
            return
        }
        FlexUtil.rejectPromise(flexWebView, funName, reason)
        finish()
    }

    fun reject() {
        if (isFinished) {
            FlexUtil.INFO(FlexException.ERROR9)
            return
        }
        FlexUtil.rejectPromise(flexWebView, funName)
        finish()
    }

    private fun finish() {
        isFinished = true
        onFinished?.invoke()
    }

}
